docs/Threading.md

```diff
@@ -0,0 +1,167 @@
+
+
+Locking
+-------
+
+There are two primitives for building locks in the lthreads implementation.
+Ticket locks are spinlocks, futexes provide a fast path using atomic operations but then sleep.
+
+### Ticked locks
+
+Ticket locks are defined and documented in [src/include/enclave/ticketlock.h](src/include/enclave/ticketlock.h).
+Acquiring a ticket lock will spin in a compare and swap (CAS) loop.
+If an lthread yields while holding a ticket lock, there is a chance of deadlock because the next thread to try to acquire the lock will spin in a CAS loop and not yield.
+This may, in turn, prevent the thread holding the lock from being able to run.
+
+Ticket locks are used to implement some parts of the lthread scheduler.
+This means that there is no way for them to be made to yield in their spin loop.
+Yielding transfers control from the current lthread to the scheduler and so the concept does not exist when the lthread is running.
+
+### Lthread futexes
+
+The lthread futex implementation is a simplified version of the Linux `futex` system call.
+A futex is a 32-bit word that can be used to hold state used to implement arbitrary locking primitives.
+On the fast path, all updates to the futex word are done by CPU atomic operations.
+For example, a mutex could be implemented by treating a value of 0 as unlocked and a value of 1 as locked, with a CAS changing the state from 0 to 1 to acquire the lock on the fast path.
+The futex calls handle the slow path, where one or more threads need to wait.
+
+The `enclave_futex_wait` call suspends the thread until a corresponding `enclave_futex_wake` call on the same key.
+The futex system is responsible for ensuring that wake-up events are not lost and so acquires a global lock and then checks that the expected value is still present at the memory location.
+This lock ensures that a `enclave_futex_wake` call either happens after the `enclave_futex_wait` has sent the calling thread to sleep or happens before and prevents the thread from sleeping.
+
+The sleeping threads are stored in a linked list, threaded through the thread pointers.
+A `enclave_futex_wait` call can return (waking up the sleeping thread) as a result of one of two things.
+Either a `enclave_futex_wake` acquires the lock, finds a waiting thread, and schedules it, or the `futex_tick` call triggers the timeout.
+
+The lthread scheduler calls `futex_tick` in between scheduling threads to wake up any sleeping threads that have timed out.
+This tries to acquire the ticket lock but returns immediately if it fails.
+This can happen only if a thread is currently doing a futex call or if another thread is handling the tick.
+
+### LKL host interface synchronisation primitives
+
+The futex code is used in [src/lkl/posix-host.c](../src/lkl/posix-host.c) to implement two synchronisation primitives for use by LKL.
```

Comment:
```suggestion
The aforementioned enclave futex code is used in [src/lkl/posix-host.c](https://github.com/lsds/sgx-lkl/blob/oe_port/src/lkl/posix-host.c) to implement two synchronisation primitives for use by LKL.
```

to match the absolutes used most other places in the document. Changing to relative for those and keeping this relative would work as well.

Reply:
See above: I wish GitHub had support for the in-document anchors in comments that the Listings package for LaTeX provides.